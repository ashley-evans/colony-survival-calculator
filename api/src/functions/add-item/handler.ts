import Ajv, { JSONSchemaType } from "ajv";
import {
    S3Client,
    GetObjectCommand,
    GetObjectCommandInput,
} from "@aws-sdk/client-s3";

import type { S3EventHandler } from "../../interfaces/S3EventHandler";

type ValidS3EventRecord = {
    s3: {
        bucket: {
            name: string;
        };
        object: {
            key: string;
        };
    };
};

const eventSchema: JSONSchemaType<ValidS3EventRecord> = {
    type: "object",
    properties: {
        s3: {
            type: "object",
            properties: {
                bucket: {
                    type: "object",
                    properties: {
                        name: { type: "string" },
                    },
                    required: ["name"],
                },
                object: {
                    type: "object",
                    properties: {
                        key: { type: "string" },
                    },
                    required: ["key"],
                },
            },
            required: ["bucket", "object"],
        },
    },
    required: ["s3"],
};

const ajv = new Ajv();
const validateEventRecord = ajv.compile<ValidS3EventRecord>(eventSchema);

const client = new S3Client({});

const handler: S3EventHandler<void> = async (event) => {
    if (!event.Records || event.Records.length === 0) {
        throw "No event records provided";
    }

    const record = event.Records[0];
    if (!record) {
        throw "";
    }

    if (validateEventRecord(record)) {
        const input: GetObjectCommandInput = {
            Key: record.s3.object.key,
            Bucket: record.s3.bucket.name,
        };

        await client.send(new GetObjectCommand(input));
    } else {
        throw validateEventRecord.errors;
    }
};

export { handler };
