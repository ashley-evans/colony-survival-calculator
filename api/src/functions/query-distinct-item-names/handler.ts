import { GraphQLEventHandler } from "../../interfaces/GraphQLEventHandler";
import { queryDistinctItemNames } from "./domain/query-distinct-item-names";

const handler: GraphQLEventHandler<never, string[]> = async () => {
    return await queryDistinctItemNames();
};

export { handler };
