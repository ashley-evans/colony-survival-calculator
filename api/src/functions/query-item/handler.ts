import type { Item, QueryItemArgs } from "../../graphql/schema";
import type { GraphQLEventHandler } from "../../interfaces/GraphQLEventHandler";
import { queryItem } from "./domain/query-item";
import { QueryFilters } from "./interfaces/query-item-primary-port";

const handler: GraphQLEventHandler<QueryItemArgs, Item[]> = async (event) => {
    const filters: QueryFilters | undefined = event.arguments.filters
        ? {
              name: event.arguments.filters.name ?? undefined,
              minimumCreators:
                  event.arguments.filters.minimumCreators ?? undefined,
              creator: event.arguments.filters.creator ?? undefined,
          }
        : undefined;

    try {
        return await queryItem(filters);
    } catch {
        throw new Error(
            "An error occurred while fetching item details, please try again."
        );
    }
};

export { handler };
