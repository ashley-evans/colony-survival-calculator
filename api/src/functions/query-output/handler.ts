import type { OutputResult, QueryOutputArgs } from "../../graphql/schema";
import type { GraphQLEventHandler } from "../../interfaces/GraphQLEventHandler";
import { calculateOutput } from "./domain/output-calculator";
import { AvailableToolsSchemaMap, OutputUnit } from "../../common";
import {
    INVALID_ITEM_NAME_ERROR,
    INVALID_WORKERS_ERROR,
    TOOL_LEVEL_ERROR_PREFIX,
    UNKNOWN_ITEM_ERROR,
} from "./domain/errors";

const exactUserErrors = new Set([
    INVALID_ITEM_NAME_ERROR,
    INVALID_WORKERS_ERROR,
    UNKNOWN_ITEM_ERROR,
]);

const isUserError = ({ message }: Error): boolean => {
    return (
        exactUserErrors.has(message) ||
        message.startsWith(TOOL_LEVEL_ERROR_PREFIX)
    );
};

const handler: GraphQLEventHandler<QueryOutputArgs, OutputResult> = async (
    event,
) => {
    const { name, workers, unit, maxAvailableTool, creator, hasMachineTools } =
        event.arguments;

    try {
        const output = await calculateOutput({
            name,
            workers,
            unit: OutputUnit[unit],
            ...(maxAvailableTool
                ? {
                      maxAvailableTool:
                          AvailableToolsSchemaMap[maxAvailableTool],
                  }
                : {}),
            ...(hasMachineTools !== undefined && hasMachineTools !== null
                ? { hasMachineTools }
                : {}),
            ...(creator ? { creator } : {}),
        });

        return { __typename: "OptimalOutput", amount: output };
    } catch (ex) {
        if (ex instanceof Error && isUserError(ex)) {
            return { __typename: "UserError", message: ex.message };
        }

        throw ex;
    }
};

export { handler };
