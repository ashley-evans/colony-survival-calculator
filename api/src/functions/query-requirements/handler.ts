import { ToolSchemaMap } from "../../common/modifiers";
import type { QueryRequirementArgs, Requirement } from "../../graphql/schema";
import type { GraphQLEventHandler } from "../../interfaces/GraphQLEventHandler";
import { queryRequirements } from "./domain/query-requirements";

const handler: GraphQLEventHandler<
    QueryRequirementArgs,
    Requirement[]
> = async (event) => {
    const { name, workers, maxAvailableTool, creatorOverrides } =
        event.arguments;

    return await queryRequirements({
        name,
        workers,
        ...(maxAvailableTool
            ? { maxAvailableTool: ToolSchemaMap[maxAvailableTool] }
            : {}),
        ...(creatorOverrides ? { creatorOverrides } : {}),
    });
};

export { handler };
