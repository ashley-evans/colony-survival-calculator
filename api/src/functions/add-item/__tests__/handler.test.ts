import type { S3EventRecord } from "aws-lambda";
import {
    S3Client,
    GetObjectCommand,
    GetObjectCommandInput,
} from "@aws-sdk/client-s3";
import { mockClient } from "aws-sdk-client-mock";

import {
    createS3Event,
    createS3EventBucketDetails,
    createS3EventBucketObjectDetails,
    createS3EventNotificationDetails,
    createS3EventRecord,
} from "../../../../test";

import { handler } from "../handler";

const mockS3Client = mockClient(S3Client);

beforeAll(() => {
    jest.spyOn(console, "log").mockImplementation(() => undefined);
});

const EXPECTED_BUCKET_NAME = "test_bucket_name";
const EXPECTED_KEY = "seeds/items.json";

function createValidEventRecord(key: string): S3EventRecord {
    return createS3EventRecord(
        createS3EventNotificationDetails(
            createS3EventBucketDetails(EXPECTED_BUCKET_NAME),
            createS3EventBucketObjectDetails(key)
        )
    );
}

const validRecord = createValidEventRecord(EXPECTED_KEY);

beforeEach(() => {
    mockS3Client.reset();
});

test.each([
    ["undefined records are provided", undefined],
    ["no records are provided", []],
    ["a record with missing notification details", [createS3EventRecord()]],
    [
        "a record with missing bucket details",
        [
            createS3EventRecord(
                createS3EventNotificationDetails(
                    undefined,
                    createS3EventBucketObjectDetails(EXPECTED_KEY)
                )
            ),
        ],
    ],
    [
        "a record with missing bucket name",
        [
            createS3EventRecord(
                createS3EventNotificationDetails(
                    createS3EventBucketDetails(),
                    createS3EventBucketObjectDetails(EXPECTED_KEY)
                )
            ),
        ],
    ],
    [
        "a record with missing bucket object details",
        [
            createS3EventRecord(
                createS3EventNotificationDetails(
                    createS3EventBucketDetails(EXPECTED_BUCKET_NAME),
                    undefined
                )
            ),
        ],
    ],
    [
        "a record with missing bucket object key",
        [
            createS3EventRecord(
                createS3EventNotificationDetails(
                    createS3EventBucketDetails(EXPECTED_BUCKET_NAME),
                    createS3EventBucketObjectDetails()
                )
            ),
        ],
    ],
])(
    "throws an error if an event with %s",
    async (_: string, records?: S3EventRecord[]) => {
        const event = createS3Event(records);

        expect.assertions(1);
        await expect(handler(event)).rejects.toMatchSnapshot();
    }
);

test("calls S3 to get item JSON details given single valid record", async () => {
    const event = createS3Event([validRecord]);

    await handler(event);

    const calls = mockS3Client.commandCalls(GetObjectCommand);
    expect(calls).toHaveLength(1);
    expect(calls[0]?.args).toHaveLength(1);

    const input = calls[0]?.args[0].input as GetObjectCommandInput;
    expect(input.Bucket).toEqual(EXPECTED_BUCKET_NAME);
    expect(input.Key).toEqual(EXPECTED_KEY);
});

test("does not call S3 if given a valid event with a unknown object key", async () => {
    const record = createValidEventRecord("unexpected");
    const event = createS3Event([record]);

    await handler(event);

    const calls = mockS3Client.commandCalls(GetObjectCommand);
    expect(calls).toHaveLength(0);
});
