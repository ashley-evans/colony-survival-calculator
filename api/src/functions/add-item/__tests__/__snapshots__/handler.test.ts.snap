// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws an error if an event with a record with missing bucket details 1`] = `
[
  {
    "instancePath": "/s3/bucket",
    "keyword": "type",
    "message": "must be object",
    "params": {
      "type": "object",
    },
    "schemaPath": "#/properties/s3/properties/bucket/type",
  },
]
`;

exports[`throws an error if an event with a record with missing bucket name 1`] = `
[
  {
    "instancePath": "/s3/bucket/name",
    "keyword": "type",
    "message": "must be string",
    "params": {
      "type": "string",
    },
    "schemaPath": "#/properties/s3/properties/bucket/properties/name/type",
  },
]
`;

exports[`throws an error if an event with a record with missing bucket object details 1`] = `
[
  {
    "instancePath": "/s3/object",
    "keyword": "type",
    "message": "must be object",
    "params": {
      "type": "object",
    },
    "schemaPath": "#/properties/s3/properties/object/type",
  },
]
`;

exports[`throws an error if an event with a record with missing bucket object key 1`] = `
[
  {
    "instancePath": "/s3/object/key",
    "keyword": "type",
    "message": "must be string",
    "params": {
      "type": "string",
    },
    "schemaPath": "#/properties/s3/properties/object/properties/key/type",
  },
]
`;

exports[`throws an error if an event with a record with missing notification details 1`] = `
[
  {
    "instancePath": "/s3",
    "keyword": "type",
    "message": "must be object",
    "params": {
      "type": "object",
    },
    "schemaPath": "#/properties/s3/type",
  },
]
`;

exports[`throws an error if an event with no records are provided 1`] = `"No event records provided"`;

exports[`throws an error if an event with undefined records are provided 1`] = `"No event records provided"`;
